### PR TITLE
REST - Channels & publishing messages

### PR DESCRIPTION
**Pull request by tiholic**
Following conditions picked from Ruby Spec and IDL, as acceptance criteria for this PR:

> Conditions to be satisfied for **\<String input>**:
> 1. can be `UTF-8` string
> 2. if given as `SHIFT_JIS` string will be converted to `UTF-8`, but is compatible with original encoding
> 3. if given as `ASCII_8BIT` string will be converted to `UTF-8`, but is compatible with original encoding

### Channels spec and Channels object spec: [IDL](https://docs.ably.io/client-lib-development-guide/features/#RSN1) | [Ruby Spec](https://github.com/ably/ably-ruby-rest/blob/main/SPEC.md#ablyrestchannels) | [Ruby Spec - 1](https://github.com/ably/ably-ruby-rest/blob/main/SPEC.md#ablyrestchannels-1)

* (RSN1) `Channels` is a collection of `Channel` objects accessible through `Rest#channels`
    - [x] Must override _List accessor operator_ and return `Channel` instance if available, otherwise create one. This should work similar to the get function as per RSN3a, but without `ChannelOptions`.
    - [x] Allows querying length
    - [x] each returns an iterator and yields a channel on each iteration.
- [x] (RSN2) Methods should exist to check if a channel exists or iterate through the existing channels
* (RSN3) `Channels#get` function:
    - [x] (RSN3a) Creates a new `Channel` object for the specified channel if none exists, or returns the existing channel. `ChannelOptions` can be provided in an optional second argument
    - [x] (RSN3b) If options are provided, the options are set on the Channel
    - [x] (RSN3c) Accessing an existing Channel with options in the form `Channels#get(channel, options)` will update the options on the channel and then return the existing Channel object. (Note that this is soft-deprecated and may be removed in a future release, so should not be implemented in new client libraries. The supported way to update a set of `ChannelOptions` is `Channel#setOptions` -- Totally depends on `ably-java` and `ably-cocoa`)
    * (RSN4) `Channels#release` function:
        - [x] (RSN4a) Takes one argument, the channel name, and releases the corresponding channel entity (that is, deletes it to allow it to be garbage collected)
        - [x] (RSN4b) Calling `release()` with a channel name that does not correspond to an extant channel entity must return without error
### `#publish` method of Channel spec : [IDL](https://docs.ably.io/client-lib-development-guide/features/#RSL1) | [Ruby Spec](https://github.com/ably/ably-ruby-rest/blob/main/SPEC.md#ablyrestchannel)

- [x] (RSL1a) Expects either a Message object, an array of Message objects, or a name string and data payload
- [x] (RSL1b) When name and data (or a Message) is provided, a single message is published to Ably
- [x] (RSL1c) When an array of Message objects is provided, a single request is made to Ably
- [x] (RSL1d) Raises an error if the message was not successfully published to Ably
- [x] (RSL1e) Allows name and/or data to be null. If any of the values are null, that property is not sent to Ably

* (RSL1m) The `Message.clientId` must be left alone (that is, it will be there if it was set in the `Message` object passed to `publish()`, else left unset). The client library must not try and set it from the client-library-wide clientId; that is achieved by `RSA7e` for basic auth or implicit in the credentials for token auth. The following tests can be used to check for correct clientId handling:
    - [x] (RSL1m1) Publishing a `Message` with no `clientId` when the `clientId` is set to some value in the client options should result in a message received with the `clientId` property set to that value
    - [x] (RSL1m2) Publishing a `Message` with a `clientId` set to the same value as the `clientId` in the client options should result in a message received with the `clientId` property set to that value
    - [x] (RSL1m3) Publishing a `Message` with a `clientId` set to a value from an unidentified client (no `clientId` in the client options and credentials that can assume any `clientId`) should result in a message received with the `clientId` property set to that value
    - [x] (RSL1m4) Publishing a `Message` with a `clientId` set to a different value from the `clientId` in the client options should result in a message being rejected by the server.

- [x] ~(RSL1h) The `publish(name, data)` form should not take any additional arguments. If a client library has supported additional arguments to the `(name, data)` form (e.g. separate arguments for clientId and extras, or a single attributes argument) in any 1.x version, it should continue to do so until version 2.0.~ (applicable for platform)
- [x] (RSL1i) If the total size of the `message` or (if publishing an array) messages, calculated per TO3l8, exceeds the `maxMessageSize`, then the client library should reject the publish and indicate an error with code `40009`
- [x] ~(RSL1j) When `Message` objects are provided, any valid `Message` attribute (that is, an attribute specified in TM2) that is supplied by the caller must be included in the encoded message. (This does not mean it must be included unaltered; for example the data and encoding will be subject to processing per RSL4)~ (applicable for platform)
- [x] without adequate permissions on the channel raises a permission error when publishing (with code `40160`. as per https://github.com/ably/ably-common/blob/main/protocol/errors.json )
- [x] With a non-ASCII channel name: stubs and correctly encodes channel name.

### Message object spec : [IDL](https://docs.ably.io/client-lib-development-guide/features/#TM1) | [Ruby Spec](https://github.com/ably/ably-ruby-rest/blob/main/SPEC.md#ablymodelsmessage)

Attributes in message:
- [x] `#id` retrieves attribute :id
- [x] `#name` retrieves attribute :name
- [x] `#client_id` retrieves attribute :client_id
- [x] `#data` retrieves attribute :data
- [x] `#encoding` retrieves attribute :encoding
- [x] `==` - is true when attributes are the same, is false when attributes are not the same, is false when class type differs
- [x] Message once created is immutable
- [x] `#timestamp` retrieves timestamp as DateTime object
- [x] `#extras` (#TM2i) Can be `null`, `map`, or `list`. Should throw exception or not allow syntactically to pass string or number, or any other types.
- [x] `#connection_id` can be `null` or a `string`

Message object can be initialized with:
- [x] `name` as "\<String input>" (defined above), or `null` but other types are not allowed
- [x] `client_id` as "\<String input>" (defined above), or `null` but other types are not allowed
- [x] `encoding` as "\<String input>" (defined above), or `null` but other types are not allowed

### Channel object spec: [Ruby Spec](https://github.com/ably/ably-ruby-rest/blob/main/SPEC.md#ablyrestchannel-1)

`channel name` argument in `constructor`
- [x] Satisfies conditions defined above for "\<String input>"
- [x] **cannot** be `null`, or any other type

`name` argument of `#publish`
- [x] Satisfies conditions defined above for "\<String input>"
- [x] `int` or any other types is not permitted

### CODE REVIEW

**Review by zoechi — CHANGES_REQUESTED** [review]

File: lib/src/impl/platform_object.dart [hunk 1]
```diff
@@ -27,45 +26,45 @@ abstract class PlatformObject {
 
   Future<int> get handle async => _handle ??= _acquireHandle();
 
-  Future<int> _acquireHandle() => createPlatformInstance().timeout(
-    _acquireHandleTimeout, onTimeout: () {
-    _handle = null;
-    throw TimeoutException('Acquiring handle timed out.', _acquireHandleTimeout);
-  }).then((value) => _handleValue=value);
+  Future<int> _acquireHandle() =>
```
[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/impl/platform_object.dart]: Resolved in previous PRs

File: lib/src/impl/realtime/channels.dart [hunk 2]
```diff
@@ -1,16 +1,17 @@
...
 
 import '../platform_object.dart';
 
-
-class RealtimePlatformChannel extends PlatformObject implements spec.RealtimeChannel{
-
+class RealtimePlatformChannel extends PlatformObject
+    implements spec.RealtimeChannel {
   @override
   spec.AblyBase ably;
```
[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/impl/realtime/channels.dart]: Resolved in earlier PRs

File: lib/src/impl/realtime/connection.dart [hunk 3]
```diff
@@ -6,21 +6,22 @@ import 'package:ably_flutter_plugin/src/impl/realtime/realtime.dart';
 import '../../spec/spec.dart' show Connection, ConnectionState, ErrorInfo;
 import '../platform_object.dart';
 
-
 class ConnectionPlatformObject extends PlatformObject implements Connection {
-
   Realtime realtimePlatformObject;
```
[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/impl/realtime/connection.dart]: addressed in 28167ab2961aa8cf82b6e20cfc14859b65078257

[the remark this answers is not in the document]
Comment (jamienewcomb) [reply lib/src/impl/realtime/connection.dart]: is this resolved now @tiholic ?

File: android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java [hunk 4]
```diff
@@ -74,39 +76,43 @@ T decode(Map<String, Object> jsonMap) {
     codecMap = new HashMap<Byte, CodecPair>() {
       {
         put(PlatformConstants.CodecTypes.ablyMessage,
-          new CodecPair<>(self::encodeAblyFlutterMessage, self::decodeAblyFlutterMessage));
+            new CodecPair<>(self::encodeAblyFlutterMessage, self::decodeAblyFlutterMessage));
```
[the remark this answers is not in the document]
Comment (tiholic) [reply android/src/main/java/io/ably/flutter/plugin/AblyMessageCodec.java:79]: I've updated indentation levels for continuation indent in Studio.

File: lib/src/codec.dart [hunk 5]
```diff
@@ -273,6 +282,46 @@ class Codec extends StandardMessageCodec {
...
+        return TxEnumConstants.publish;
+      case ChannelMode.subscribe:
+        return TxEnumConstants.subscribe;
+      case ChannelMode.presenceSubscribe:
+        return TxEnumConstants.presenceSubscribe;
+    }
+    return null;
+  }
+
+  /// Encodes [TokenRequest] to a Map
```
[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/codec.dart]: suggestion accepted: a1fd1ea18fe7f5f80ba51da6a67d776fe2447a63

File: lib/src/codec.dart [hunk 6]
```diff
@@ -273,6 +282,46 @@ class Codec extends StandardMessageCodec {
     return jsonMap;
   }
 
+  /// Encodes [TokenRequest] to a Map
```
[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/codec.dart]: Suggestion accepted: 8351bb8a73923027a0e794236a920f6dcf6a9289

File: lib/src/codec.dart [hunk 7]
```diff
@@ -357,6 +406,18 @@ class Codec extends StandardMessageCodec {
     return jsonMap;
   }
 
+  Map<String, dynamic> _encodeErrorInfo(final ErrorInfo v) {
```
[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/codec.dart:413]: added commentary on all methods: bd6298413ec66a454c8f8dc19e5c41c7bd0c851c

_Ignored it on purpose earlier as private methods won't be included in docs._

File: lib/src/spec/common.dart [hunk 8]
```diff
@@ -901,35 +907,65 @@ class Stats {
   StatsRequestCount tokenRequests;
 }
 
+/// Iterator class for [Channels.iterator]
+class _ChannelIterator<T> implements Iterator<T> {
```
[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/spec/common.dart:911]: Yes, first time I wrote a custom iterator. Expected it to be simpler!

[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/spec/common.dart:911]: Yes, first time I wrote a custom iterator in dart. Expected it to be simpler!

File: ios/Classes/AblyFlutterPlugin.m [hunk 9]
```diff
@@ -253,12 +275,36 @@ -(void)registerWithCompletionHandler:(FlutterResult)completionHandler;
...
+    AblyFlutterMessage *const data = message.message;
+    NSNumber *const realtimeHandle = data.handle;
+    ARTRealtime *const realtimeWithHandle = [ably realtimeWithHandle: realtimeHandle];
+    
+    NSDictionary *const realtimePayload = data.message;
+    NSString *const channelName = (NSString*)[realtimePayload objectForKey:TxTransportKeys_channelName];
+    ARTRealtimeChannelOptions *const channelOptions = (ARTRealtimeChannelOptions*)[realtimePayload objectForKey:TxTransportKeys_options];
+
+    ARTRealtimeChannel *const channel = [realtimeWithHandle.channels get:channelName];
+    const id cbk = ^(ARTPaginatedResult<ARTMessage *> * _Nullable paginatedResult, ARTErrorInfo * _Nullable error) {
```
[the remark this answers is not in the document]
Comment (tiholic) [reply ios/Classes/AblyFlutterPlugin.m]: renamed: d552f096117318de9e721db3d739769baf841f4a

File: ios/Classes/codec/AblyPlatformConstants.m [hunk 10]
```diff
@@ -108,6 +112,16 @@ @implementation TxClientOptions
 NSString *const TxClientOptions_transportParams = @"transportParams";
 @end
 
+@implementation TxRestChannelOptions
```
[the remark this answers is not in the document]
Comment (tiholic) [reply ios/Classes/codec/AblyPlatformConstants.m]: addressed: 7179b2dd6a9f8e4688d671d7f47b2a43284c21b8

File: ios/Classes/codec/AblyPlatformConstants.h [hunk 11]
```diff
@@ -128,6 +134,16 @@ extern NSString *const TxClientOptions_channelRetryTimeout;
 extern NSString *const TxClientOptions_transportParams;
 @end
 
+@interface TxRestChannelOptions : NSObject
```
[the remark this answers is not in the document]
Comment (tiholic) [reply ios/Classes/codec/AblyPlatformConstants.h]: addressed: 7179b2dd6a9f8e4688d671d7f47b2a43284c21b8

File: lib/src/spec/common.dart [hunk 12]
```diff
@@ -901,35 +907,65 @@ class Stats {
...
   /// stores channel name vs instance of [ChannelType]
   final _channels = <String, ChannelType>{};
 
   /// creates a channel with provided name and options
-  ChannelType createChannel(String name, ChannelOptions options);
-
-  /// returns all channels
-  Iterable<ChannelType> get all => _channels.values;
+  ///
+  /// This is a private method to be overridden by implementation classes
```
[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/spec/common.dart:942]: addressed: 3166620f89e56a28538514b2b02da9e589305283

File: lib/src/impl/realtime/connection.dart (hunk 3 above)
[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/impl/realtime/connection.dart]: @jamienewcomb yes

File: lib/src/spec/message.dart [hunk 13]
```diff
@@ -49,9 +49,46 @@ class MessageData<T> {
...
+    );
+    if (value is MessageExtras) {
+      return value;
+    } else if (value is Map) {
+      return MessageExtras<Map>(value);
+    } else if (value is List) {
+      return MessageExtras<List>(value);
+    } else {
+      throw AssertionError(
+        'Message extras must be either `Map`, `List` or `null`.'
```
[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/spec/message.dart]: fixed: 246e6b6ceb9a951348e9d169348905faf08fe4f4

[the remark this answers is not in the document]
Comment (tiholic) [reply lib/src/spec/message.dart]: Further optimization of `MessageExtras` is being handled here: https://github.com/ably/ably-flutter/pull/123

**Review by QuintinWillison — APPROVED** [review]
I've not gone through line by line, though I have checked the commits you've submitted since my last review in order to resolve those conversations.

The `main` branch was out of date so I've boshed GitHub's update button, resolving that with the merge commit that is https://github.com/ably/ably-flutter/pull/45/commits/fa34b4340124032984fc03e0343ce5be4828b496.

Once checks pass then I'm happy to land this.

**Pull request merged**